### ISSUE TITLE
Gatsby: WebpackError: ReferenceError: window is not defined

### ISSUE THREAD

**Issue by Hahlh** [Gherciu/react-p5#47]
Although there is a check in place:
https://github.com/Gherciu/react-p5/blob/f029ba722330d06406bce99fad503ff151adfc53/src/index.js#L4

The build still fails for me. This could be the case, because of the one-liner if statement. It could lead to an error during static analysis, but that's speculation. Do you have more insight on what could cause this even though it should normally be prevented? 

The complete error from the console:
```node
WebpackError: ReferenceError: window is not defined
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:330179
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1439
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1475
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:272442
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1439
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1475
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:230003
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1439
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1570
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1591
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:612542
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:373
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:612637
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:373
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1176
  
  - index.js:1 
    [Dolce Stil Criollo]/[react-p5]/build/index.js:1:1193
```

Could the code be included unminified during development and be minified only for production? Are the gains through minifing so great here, that it justifies the increased difficulty with debugging? 
Otherwise it makes all errors really hard to follow, also with errors unrelated to the build.

I will create a PR for adding brackets to the one-liner as a fail-safe. I am not sure whether that will actually fix the issue, but it is worth a shot? Unfortunately I can't test this out myself as there are some build steps involved that I don't quite understand. Maybe you could clarify in the docs, how one goes about local development of this package.

I will also create a new issue asking about local development of this package.

Is anybody else building successfully using Gatsby v3?

**Comment by Hahlh** [comment]
Unfortunately the change doesn't seem to have worked.

Do you use build-time SSR and can successfully deploy?
I will be able to test changes to the code as soon as I am able to develop `react-p5` locally :+1:

**Comment by Gherciu** [comment]
@Hahlh  I think a workaround on that is to just not render canvas or load p5 in SSR mode because it doesn't make sense....it doesn't affect SEO or anything else if we will not render it in SSR

**Comment by Gherciu** [comment]
because the problem here is actually with p5 itself but not with (`react-p5` it just renders a <div/> in SSR mode).....but `p5` needs access to `window` etc.....that's why it throws errors

**Comment by Hahlh** [comment]
You would normally implement type checks in your code that prevent errors during build:

Something like:
```javascript
if (typeof window !== 'undefined') {
    window.doSomething
}
```

In this case the error occurs directly in `react-p5`, meaning that I can't do anything about it without changing `react-p5`:
![image](https://user-images.githubusercontent.com/45669613/113316977-33ee9b80-930f-11eb-8270-fd888bc04639.png)

![image](https://user-images.githubusercontent.com/45669613/113317050-45d03e80-930f-11eb-98be-a71c780dd13d.png)



I am curious as surely other people must have it running with Gatsby?
Maybe @lambertbrady can weigh in as he proposed the first change to the window check in `react-p5`?

Did it work for you after the change?
Did you additionally have to wrap all the `react-p5` related code in a window-checker or was wrapping those directly referring to `window` enough?

**Comment by Hahlh** [comment]
Update:

If I comment all `react-p5` related code out, it builds again successfully. So something in my code is triggering `react-p5` and it in turn fails, because `window` is undefined.

That should make the question to you, @lambertbrady, solved. Please excuse the unnecessary ping.

**Comment by Hahlh** [comment]
As a side note:

This is what the error logged to the terminal looks like in this case:
![image](https://user-images.githubusercontent.com/45669613/113318838-263a1580-9311-11eb-8af6-99bb19d1892f.png)

I am not sure whether that is something that can be improved in `react-p5`?

**Comment by Hahlh** [comment]
Another update:

If I leave all the code unchanged and only comment out the Sketch component return, the code builds successfully. 

So that should mean, that additional `window` checks are required in `react-p5` to prevent calls to `p5`, that result in a `window undefined`?

**Comment by Gherciu** [comment]
@Hahlh as a workaround use  dinamyc imports ....just to import this library when window is available othervise no. Example.
```js
if(window) {
const Sketch = require(react-p5)
}
```

**Comment by Hahlh** [comment]
@Gherciu 

I hope you are doing great! 

```javascript
let Sketch

if (typeof window !== "undefined") {
  Sketch = require("react-p5")
}
```
works, although it's means mixing `import` and `require` and although the import should normally be top level?

The problem with this is, that it also somehow changes my setup, meaning where the canvas is placed etc.

Expected result:
![image](https://user-images.githubusercontent.com/45669613/113332733-dc0d6000-9321-11eb-8ae6-2cf38cc961f7.png)

Actual result:
![image](https://user-images.githubusercontent.com/45669613/113332082-03176200-9321-11eb-9384-f66f03e98b95.png)

I will now explore two alternative approaches from the Gatsby docs.
[Customizing the webpack config to use a dummy module during build and using loadable-components.](https://www.gatsbyjs.com/docs/debugging-html-builds/#fixing-third-party-modules)

If those don't succeed, I may try to force the setup to happen later with a useEffect, although I am not sure whether that will introduce scoping problems.

Thank you again for the fast help, @Gherciu !

**Comment by Hahlh** [comment]
In the example above, using dynamic imports as you suggested, and in using a custom webpack config, the canvas seems to get attached to the wrong parent.

In this case the absolute div on the right side. 

Could I do something about that or specify precisely what the parent element should be?
